# **SVM**

In [ ]:
import os
import pandas as pd
data = pd.read_csv('/content/extracted_features (1).csv')
data

filename       label  bow_abandon  \
0        preprocessed_source-document00014.txt      source            1   
1        preprocessed_source-document00039.txt      source            1   
2        preprocessed_source-document00022.txt      source            1   
3        preprocessed_source-document00088.txt      source            0   
4        preprocessed_source-document00041.txt      source            0   
..                                         ...         ...          ...   
187  preprocessed_suspicious-document00050.txt  suspicious            0   
188  preprocessed_suspicious-document00052.txt  suspicious            0   
189  preprocessed_suspicious-document00034.txt  suspicious            0   
190  preprocessed_suspicious-document00008.txt  suspicious            0   
191  preprocessed_suspicious-document00046.txt  suspicious            0   

     bow_abandoned  bow_ability  bow_able  bow_abner  bow_abode  bow_about  \
0                0            1         1          0          0         12   
1                0            0         3          0          0         16   
2                1            3         9          0          0         44   
3                2            0         3          0          0         12   
4                0            2         0          0          0          2   
..             ...          ...       ...        ...        ...        ...   
187              0            0         3          0          0         31   
188              0            0         1          0          0          0   
189              0            0         0          0          0          0   
190              1            3        27        295          0        157   
191              0            0         1          0          0          2   

     bow_above  ...    w2v_90    w2v_91    w2v_92    w2v_93    w2v_94  \
0            2  ...  1.287014 -0.252784 -0.001455  0.409457  0.797398   
1            1  ...  0.992542 -0.202871  0.096220  0.321591  0.920017   
2           26  ...  0.694806 -0.208175  0.098772  0.142081  0.717311   
3            6  ...  0.943576 -0.226208  0.141268  0.163794  0.858136   
4            2  ...  0.763065 -0.180428  0.084040  0.127569  0.799522   
..         ...  ...       ...       ...       ...       ...       ...   
187          2  ...  1.326246 -0.257754 -0.048474  0.487143  0.800616   
188          0  ...  1.015267 -0.184723  0.138802  0.286528  0.880048   
189          0  ...  1.132809 -0.296026  0.016198  0.325406  0.874679   
190         14  ...  1.194706 -0.232931  0.005586  0.386601  0.751338   
191          5  ...  1.180322 -0.202994  0.084947  0.417208  0.859609   

       w2v_95    w2v_96    w2v_97    w2v_98    w2v_99  
0    0.207746 -0.169751 -0.089558  0.231392 -0.399079  
1    0.243611 -0.299805 -0.217917  0.053264 -0.240799  
2    0.154940 -0.358141 -0.327016 -0.063200 -0.082171  
3    0.161049 -0.430236 -0.134742 -0.031938 -0.263049  
4    0.164538 -0.500480 -0.291192 -0.088641 -0.044195  
..        ...       ...       ...       ...       ...  
187  0.210270 -0.077548 -0.078374  0.295074 -0.368793  
188  0.147166 -0.383990 -0.152979 -0.049142 -0.257716  
189  0.189657 -0.255520 -0.124931  0.177988 -0.351315  
190  0.188866 -0.148061 -0.083822  0.211922 -0.348725  
191  0.225728 -0.290974 -0.174907  0.067587 -0.318920  

[192 rows x 10102 columns]

## kernel rbf 1 -- tanpa hyperparam tunning

In [ ]:
#SPLIT DATA 70:30
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
import pandas

# Load fitur yang sudah disimpan
df = pd.read_csv('extracted_features (1).csv')

X = df.drop(columns=['filename', 'label'])
y = df['label'].map({'source': 0, 'suspicious': 1})  # Contoh label encoding binary

# Split data train-test
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42, stratify=y)

# Scaling fitur (penting untuk Word2Vec dan BERT)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Inisialisasi model SVM (kernel bisa dicoba 'linear', 'rbf', dll)
svm_model = SVC(kernel='rbf', C=1, gamma='scale', random_state=42)

# Latih model
svm_model.fit(X_train_scaled, y_train)

# Prediksi pada test set
y_pred = svm_model.predict(X_test_scaled)

# Evaluasi
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy SVM: {accuracy * 100:.2f}%")

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))


Accuracy SVM: 60.34%

Classification Report:
              precision    recall  f1-score   support

           0       0.57      0.71      0.63        28
           1       0.65      0.50      0.57        30

    accuracy                           0.60        58
   macro avg       0.61      0.61      0.60        58
weighted avg       0.61      0.60      0.60        58


Confusion Matrix:
[[20  8]
 [15 15]]


In [ ]:
#SPLIT DATA 80:20
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Load fitur yang sudah disimpan
df = pd.read_csv('extracted_features (1).csv')

X = df.drop(columns=['filename', 'label'])
y = df['label'].map({'source': 0, 'suspicious': 1})  # Contoh label encoding binary

# Split data train-test
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y)

# Scaling fitur (penting untuk Word2Vec dan BERT)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Inisialisasi model SVM (kernel bisa dicoba 'linear', 'rbf', dll)
svm_model = SVC(kernel='rbf', C=1, gamma='scale', random_state=42)

# Latih model
svm_model.fit(X_train_scaled, y_train)

# Prediksi pada test set
y_pred = svm_model.predict(X_test_scaled)

# Evaluasi
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy SVM: {accuracy * 100:.2f}%")

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))


Accuracy SVM: 56.41%

Classification Report:
              precision    recall  f1-score   support

           0       0.56      0.53      0.54        19
           1       0.57      0.60      0.59        20

    accuracy                           0.56        39
   macro avg       0.56      0.56      0.56        39
weighted avg       0.56      0.56      0.56        39


Confusion Matrix:
[[10  9]
 [ 8 12]]


## Pipeline

In [7]:
# from sklearn.preprocessing import StandardScaler
# scaler = StandardScaler()
# X_train_scaled = scaler.fit_transform(X_train)
# X_test_scaled = scaler.transform(X_test)


# svm = SVC(class_weight='balanced')


# param_grid = {
#     'C': [0.1, 1, 10, 100, 500],
#     'kernel': ['linear', 'rbf', 'poly', 'sigmoid'],
#     'gamma': ['scale', 'auto', 0.01, 0.001, 0.0001],
#     'degree': [2, 3, 4]  # khusus poly
# }


# from sklearn.model_selection import StratifiedKFold

# cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
# grid_search = GridSearchCV(svm, param_grid, cv=cv, scoring='accuracy', n_jobs=-1, verbose=2)

# from sklearn.pipeline import Pipeline

# pipeline = Pipeline([
#     ('scaler', StandardScaler()),
#     ('svm', SVC(class_weight='balanced'))
# ])

# param_grid = {
#     'svm__C': [0.1, 1, 10, 100, 500],
#     'svm__kernel': ['linear', 'rbf', 'poly', 'sigmoid'],
#     'svm__gamma': ['scale', 'auto', 0.01, 0.001, 0.0001],
#     'svm__degree': [2, 3, 4]
# }

# grid_search = GridSearchCV(pipeline, param_grid, cv=cv, scoring='accuracy', n_jobs=-1, verbose=2)
# grid_search.fit(X_train, y_train)

# print("Best parameters found:", grid_search.best_params_)
# print("Best cross-validation accuracy:", grid_search.best_score_)

# y_pred = grid_search.predict(X_test)
# print("Test set accuracy:", accuracy_score(y_test, y_pred))
# print(classification_report(y_test, y_pred))


In [9]:
# SPLIT DATA 70:30
from sklearn.pipeline import Pipeline
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV, train_test_split, StratifiedKFold
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, f1_score, classification_report, confusion_matrix

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, stratify=y, random_state=42
)

pipeline = Pipeline([
    ('scaler', StandardScaler(with_mean=False)),
    ('svm', SVC(probability=True))
])

# GridSearchCV hyperparameter
param_grid = {
    'svm__C': [0.1, 1, 10, 100],
    'svm__kernel': ['linear', 'rbf', 'poly'],
    'svm__gamma': ['scale', 'auto', 0.01, 0.001],
    'svm__degree': [2, 3],
}

cv_strategy = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

grid_search = GridSearchCV(
    pipeline,
    param_grid=param_grid,
    cv=cv_strategy,
    scoring='accuracy',
    n_jobs=-1,
    verbose=2
)

# Train
grid_search.fit(X_train, y_train)

# Hasil terbaik
print("Best Parameters:", grid_search.best_params_)
print("Best CV Accuracy: {:.2f}%".format(grid_search.best_score_ * 100))

# Evaluasi pada test set
y_pred = grid_search.predict(X_test)
test_acc = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred, average='weighted')

print("\n=== EVALUATION ON TEST SET ===")
print("Test Accuracy: {:.2f}%".format(test_acc * 100))
print("F1 Score: {:.4f}".format(f1))
print("Classification Report:\n", classification_report(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))


Fitting 5 folds for each of 96 candidates, totalling 480 fits
Best Parameters: {'svm__C': 0.1, 'svm__degree': 2, 'svm__gamma': 'scale', 'svm__kernel': 'linear'}
Best CV Accuracy: 72.36%

=== EVALUATION ON TEST SET ===
Test Accuracy: 63.79%
F1 Score: 0.6380
Classification Report:
               precision    recall  f1-score   support

           0       0.62      0.64      0.63        28
           1       0.66      0.63      0.64        30

    accuracy                           0.64        58
   macro avg       0.64      0.64      0.64        58
weighted avg       0.64      0.64      0.64        58

Confusion Matrix:
 [[18 10]
 [11 19]]


In [10]:
# SPLIT DATA 80:20
from sklearn.pipeline import Pipeline
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV, train_test_split, StratifiedKFold
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, f1_score, classification_report, confusion_matrix

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)

pipeline = Pipeline([
    ('scaler', StandardScaler(with_mean=False)),
    ('svm', SVC(probability=True))
])

# GridSearchCV hyperparameter
param_grid = {
    'svm__C': [0.1, 1, 10, 100],
    'svm__kernel': ['linear', 'rbf', 'poly'],
    'svm__gamma': ['scale', 'auto', 0.01, 0.001],
    'svm__degree': [2, 3],
}

cv_strategy = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

grid_search = GridSearchCV(
    pipeline,
    param_grid=param_grid,
    cv=cv_strategy,
    scoring='accuracy',
    n_jobs=-1,
    verbose=2
)

# Train
grid_search.fit(X_train, y_train)

# Hasil terbaik
print("Best Parameters:", grid_search.best_params_)
print("Best CV Accuracy: {:.2f}%".format(grid_search.best_score_ * 100))

# Evaluasi pada test set
y_pred = grid_search.predict(X_test)
test_acc = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred, average='weighted')

print("\n=== EVALUATION ON TEST SET ===")
print("Test Accuracy: {:.2f}%".format(test_acc * 100))
print("F1 Score: {:.4f}".format(f1))
print("Classification Report:\n", classification_report(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))


Fitting 5 folds for each of 96 candidates, totalling 480 fits
Best Parameters: {'svm__C': 0.1, 'svm__degree': 2, 'svm__gamma': 'scale', 'svm__kernel': 'linear'}
Best CV Accuracy: 72.65%

=== EVALUATION ON TEST SET ===
Test Accuracy: 64.10%
F1 Score: 0.6410
Classification Report:
               precision    recall  f1-score   support

           0       0.63      0.63      0.63        19
           1       0.65      0.65      0.65        20

    accuracy                           0.64        39
   macro avg       0.64      0.64      0.64        39
weighted avg       0.64      0.64      0.64        39

Confusion Matrix:
 [[12  7]
 [ 7 13]]


## Tanpa Pipeline

In [ ]:
#SPLIT DATA 70:30

from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.metrics import classification_report, accuracy_score
import pandas as pd
import numpy as np


#load dataset fitur dan label
df = pd.read_csv('extracted_features (1).csv')


X = df.drop(columns=['filename', 'label'])
y = df['label'].apply(lambda x: 1 if x == 'suspicious' else 0)  # Contoh binari

# Bagi data train-test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)

# Inisialisasi model SVM
svm = SVC()

# Grid parameter tuning
param_grid = {
    'C': [0.1, 1, 10, 100],
    'kernel': ['linear', 'rbf', 'poly'],
    'gamma': ['scale', 'auto'],
    'degree': [2, 3, 4]  # hanya untuk 'poly'
}

# GridSearch dengan 5-fold cross-validation
grid_search = GridSearchCV(svm, param_grid, cv=5, scoring='accuracy', n_jobs=-1, verbose=2)

# Jalankan pencarian parameter terbaik
grid_search.fit(X_train, y_train)

print("Best parameters found:", grid_search.best_params_)
print("Best cross-validation accuracy:", grid_search.best_score_)

# Evaluasi model terbaik pada test set
best_svm = grid_search.best_estimator_
y_pred = best_svm.predict(X_test)

print("Test set accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))


Fitting 5 folds for each of 72 candidates, totalling 360 fits
Best parameters found: {'C': 0.1, 'degree': 2, 'gamma': 'scale', 'kernel': 'linear'}
Best cross-validation accuracy: 0.649002849002849
Test set accuracy: 0.6724137931034483
              precision    recall  f1-score   support

           0       0.66      0.68      0.67        28
           1       0.69      0.67      0.68        30

    accuracy                           0.67        58
   macro avg       0.67      0.67      0.67        58
weighted avg       0.67      0.67      0.67        58



In [ ]:
# SPLIT DATA 80:20

from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.metrics import classification_report, accuracy_score
import pandas as pd
import numpy as np

# Contoh load dataset fitur dan label
df = pd.read_csv('extracted_features (1).csv')


X = df.drop(columns=['filename', 'label'])
y = df['label'].apply(lambda x: 1 if x == 'suspicious' else 0)  # Contoh binari

# Bagi data train-test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Inisialisasi model SVM tanpa parameter khusus
svm = SVC()

# Grid parameter tuning
param_grid = {
    'C': [0.1, 1, 10, 100],
    'kernel': ['linear', 'rbf', 'poly'],
    'gamma': ['scale', 'auto'],
    'degree': [2, 3, 4]  # hanya untuk 'poly'
}

# GridSearch dengan 5-fold cross-validation
grid_search = GridSearchCV(svm, param_grid, cv=5, scoring='accuracy', n_jobs=-1, verbose=2)

# Jalankan pencarian parameter terbaik
grid_search.fit(X_train, y_train)

print("Best parameters found:", grid_search.best_params_)
print("Best cross-validation accuracy:", grid_search.best_score_)

# Evaluasi model terbaik pada test set
best_svm = grid_search.best_estimator_
y_pred = best_svm.predict(X_test)

print("Test set accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))


Fitting 5 folds for each of 72 candidates, totalling 360 fits
Best parameters found: {'C': 0.1, 'degree': 2, 'gamma': 'scale', 'kernel': 'linear'}
Best cross-validation accuracy: 0.6133333333333334
Test set accuracy: 0.6923076923076923
              precision    recall  f1-score   support

           0       0.67      0.74      0.70        19
           1       0.72      0.65      0.68        20

    accuracy                           0.69        39
   macro avg       0.69      0.69      0.69        39
weighted avg       0.70      0.69      0.69        39



### Penjelasan
Terbaik Akurasinya? Percobaan 6?

Data dibagi menjadi 80% pelatihan dan 20% pengujian. Model terbaik: C=0.1, kernel=linear. Akurasi rata-rata cross-validation (train set): 61.33%. Akurasi pada data uji: 69.23%, lebih tinggi dibanding saat cross-validation (ada kemungkinan overfitting kecil pada train).


Akurasi total: 69.23%

F1-score rata-rata (macro): 0.69

Hasilnya cukup seimbang antara kelas 0 dan 1.


### Kesimpulan

- Model SVC dengan kernel linear dan C kecil (C=0.1) memberikan performa terbaik di dataset ini.

- Akurasi uji lebih tinggi dari cross-validation, tapi model masih di bawah 70%

- Distribusi performa di kedua kelas seimbang, menunjukkan model tidak bias terlalu kuat ke salah satu kelas.